### ISSUE TITLE
"dangerous use of global this object" with goog.bind

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#834]
Using new type inference `--new_type_inf` get message `WARNING - dangerous use of the global this object` although `this` is within a `goog.bind()` that is specifying the value for `this`.

code snippet (from constructor for a class)
```javascript
  goog.asserts.setErrorHandler(
    goog.bind(
      function(/** goog.asserts.AssertionError*/a){
        window.console.log(a + " stack= "+a.stack);
        this.labTimer_.stop();
        throw a;
      } ,this)
  );
```

error message:
```
src/lab/app/SimRunner.js:114: WARNING - dangerous use of the global this object
        this.labTimer_.stop();
        ^
```

compiler version:
```
Closure Compiler (http://github.com/google/closure-compiler)
Version: v20150126-82-g85da5b6
Built on: 2015/02/11 20:55
```

**Comment by dimvar** [comment]
We have this bug also with the native Function.prototype.bind

**Comment by myphysicslab** [comment]
This is also happening with `goog.array.forEach`.  Example:

```javascript
    goog.array.forEach(this.collisions_, function(c) {
      // update the collisions to see new velocity
      c.updateCollision(this.sim_.getTime());
    }, this);
```

with NTI get this warning:

```
src/lab/engine2D/CollisionAdvance3.js:559: WARNING - dangerous use of the global this object
      c.updateCollision(this.sim_.getTime());
                        ^
```


Closure Compiler (http://github.com/google/closure-compiler)
Version: v20150505-58-g39bb337
Built on: 2015/05/21 12:45

**Comment by dimvar** [comment]
Bumping the priority since you're getting this error often.

**Comment by dimvar** [comment]
I have a pending fix for this for the .bind case. I don't know if we'll fix the goog.array.forEach case.

OTI seems to be processing some functions in weird order in order to support this, and I'm not sure if it's worth it. I'll file a separate issue for that (with lower priority) and close this issue once the .bind fix lands.

**Comment by myphysicslab** [comment]
FYI the "dangerous use of global this" warnings disappear when I remove the goog.scope in test case I'm making for #992.

**Comment by myphysicslab** [comment]
Disregard my last comment... it was due to a typo.

**Comment by myphysicslab** [comment]
Still getting the error on `goog.bind`
```
src/lab/app/SimRunner.js:166: WARNING - dangerous use of the global this object
        this.clock_.pause();
        ^
```
Here is the code:
```javascript
  goog.asserts.setErrorHandler(
    goog.bind(
      function(/** !goog.asserts.AssertionError*/a){
        window.console.log(a + " stack= "+a.stack);
        this.clock_.pause();
        throw a;
      } ,this)
  );
```
Compiler Version: v20150505-110-g5588adc

**Comment by dimvar** [comment]
In this example, the argument function has a type annotation, so NTI won't try to infer anything about its signature, including the type of <code>this</code>. You can either remove the type annotation for <code>a</code>, or declare the type of <code>this</code>.

OTI mixes and matches here to create a more complete signature, but that seems hacky to me. We try to grab types from the context only when the bound function has no type annotations.

**Comment by myphysicslab** [comment]
Remove the annotation, still get the error:
```
src/lab/app/SimRunner.js:166: WARNING - dangerous use of the global this object
        this.clock_.pause();
        ^
```
with
```javascript
  goog.asserts.setErrorHandler(
    goog.bind(
      function(a){
        window.console.log(a + " stack= "+a.stack);
        this.clock_.pause();
        throw a;
      },this)
  );
```
BTW, that code is at the end of the constructor for the class.

Same error also when annotating `this`
```javascript
  goog.asserts.setErrorHandler(
    goog.bind(
      function(/** !goog.asserts.AssertionError*/a){
        window.console.log(a + " stack= "+a.stack);
        this.clock_.pause();
        throw a;
      }, /** !myphysicslab.lab.app.SimRunner*/this)
  );
```
The following version produces no error:
```javascript
  /** 
  @type {function(this:myphysicslab.lab.app.SimRunner, !goog.asserts.AssertionError)}
  */
  var f = function(a){
        window.console.log(a + " stack= "+a.stack);
        this.clock_.pause();
        throw a;
      };
  goog.asserts.setErrorHandler(
    goog.bind(f, this)
  );
```
Similarly the following also gives no error:
```javascript
/**
* @constructor
*/
myphysicslab.lab.app.SimRunner = function(advance) {

  goog.asserts.setErrorHandler(
      goog.bind(this.handleError, this)
    );
};

/**
@param {!goog.asserts.AssertionError} a
*/
SimRunner.prototype.handleError = function(a) {
  window.console.log(a + " stack= "+a.stack);
  this.clock_.pause();
  throw a;
};
```

**Comment by dimvar** [comment]
OK, I see the issue. NTI couldn't infer the type of THIS (the 2nd argument to bind) and thus it couldn't propagate that to the bound function. Fix pending.

**Comment by myphysicslab** [comment]
confirmed that the `goog.bind` case is working in Version: v20150609-23-g19613b6
